pubsub/server.go

```diff
@@ -0,0 +1,161 @@
+// Copyright (C) 2019-2022, Ava Labs, Inc. All rights reserved.
+// See the file LICENSE for licensing terms.
+
+package pubsub
+
+import (
+	"context"
+	"net/http"
+	"sync"
+	"sync/atomic"
+	"time"
+
+	"github.com/gorilla/websocket"
+
+	"go.uber.org/zap"
+
+	"github.com/ava-labs/avalanchego/utils/logging"
+)
+
+var upgrader = websocket.Upgrader{
+	CheckOrigin: func(*http.Request) bool {
+		return true
+	},
+}
+
+// Server maintains the set of active clients and sends messages to the clients.
+//
+// Connect to the server after starting using websocket.DefaultDialer.Dial().
+type Server struct {
+	// The http server
+	s *http.Server
+	// The address to listen on
+	addr string
+	log  logging.Logger
+	lock sync.RWMutex
+	// conns a set of all our connections
+	conns *Connections
+	// Callback function when server receives a message
+	rCallback Callback
+	// Size of the ws read buffer
+	readBufferSize int
```

Comment:
You should put all of these in a `Config` struct instead of storing them directly in the struct. This also makes passing to `New` much easier.

Do something like: https://github.com/ava-labs/hypersdk/blob/a73396d63b87501b8423fff0187ebcb69d88ba07/pebble/db.go#L34-L53